Benchmarks/CppBenchmarks/src/BinaryHeap.m

```diff
@@ -0,0 +1,76 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+#import <Foundation/Foundation.h>
+
+#import "BinaryHeap.h"
+
+@implementation BinaryHeap
+{
+  CFBinaryHeapRef _storage;
+}
+
+static const void *HeapRetain(CFAllocatorRef allocator, const void *object)
+{
+  return CFRetain(object);
+}
+
+static void HeapRelease(CFAllocatorRef allocator, const void *object)
+{
+  CFRelease(object);
+}
+
+static CFComparisonResult HeapCompare(const void *lhs, const void *rhs, void *context)
+{
+  return (CFComparisonResult)[(__bridge id)lhs compare:(__bridge id)rhs];
+}
+
+- (instancetype)init
+{
+  if ((self = [super init]))
+  {
+    CFBinaryHeapCallBacks callbacks = (CFBinaryHeapCallBacks){
+      .version = 0,
+      .retain = &HeapRetain,
+      .release = &HeapRelease,
+      .copyDescription = &CFCopyDescription,
+      .compare = &HeapCompare
+    };
+    _storage = CFBinaryHeapCreate(kCFAllocatorDefault, 0, &callbacks, NULL);
+  }
+
+  return self;
+}
+
+- (void)dealloc
+{
+  CFRelease(_storage);
+}
+
+- (NSUInteger)count
+{
+  return CFBinaryHeapGetCount(_storage);
+}
+
+- (void)insert:(NSInteger)value
+{
+  NSNumber *val = @(value);
```

Comment:
To make the benchmark a fairer comparison, let's insert the `NSInteger`s directly into the `CFBinaryHeap` (rather than `NSNumber`s). All we'd need to do is update the retain/release hooks to be no-ops and update insert/remove/compare to do the appropriate pointer casting.

As it stands the `objc_msgSend` overhead is probably convoluting the benchmark.

Reply:
Wouldn't we need to manually manage the memory for the integers, as the address of `value` is only valid for the scope of the method?

Reply:
(ping, @kylemacomber)

Reply:
A primitive type like `NSInteger` is passed by value—there's no heap allocation that needs to be managed. CF collection types support elements that are the size of a pointer. `NSInteger` is the exact same size as a pointer, so we can insert it into the CFBinaryHeap and just implement retain/release to be no-ops (since, again, there's no heap allocation that needs to be reference counted).

Reply:
Ah! Got it — cast the integer to a pointer when inserting and back again when getting the minimum value/doing comparisons. I had assumed we would need to insert a pointer to the integer into the heap (hence my question above), so I started `malloc`/`free`ing.